app/models/user.rb

```diff
@@ -35,9 +35,15 @@ class User < ActiveRecord::Base
   include Facebookeable
 
   validates :username, uniqueness: true, allow_blank: true, allow_nil: true
+  validates_uniqueness_of :email, allow_blank: true, allow_nil: true
+  validates_presence_of :password unless :facebook_id.present?
+  validates :facebook_id, uniqueness: true unless :password.present?
 
-  def full_name
-    return username unless first_name.present?
-    "#{first_name} #{last_name}"
+  def to_s
+    return username
   end
+
+  def full_name
+    return username if username.present? else "#{first_name} #{last_name}"
```

Comment:
I prefer the shorthand if, because it is clearer: `username.present? ? username : "#{first_name} #{last_name}"`